File: papermerge/core/features/auth/scopes.py
"""
Permission scopes
"""
from enum import Enum
from typing import Dict, List, Set


class ScopeCategory(Enum):
    """Categories of permission scopes."""
    NODE = "node"
    DOCUMENT = "document"
    TAG = "tag"
    USER = "user"
    GROUP = "group"
    ROLE = "role"
    TASK = "task"
    OCR = "ocr"
    PAGE = "page"
    SHARED_NODE = "shared_node"
    PORTAL = "portal"
    COMMENT = "comment"


class Action(Enum):
    """Available actions for scopes."""
    CREATE = "create"
    VIEW = "view"
    UPDATE = "update"
    DELETE = "delete"
    MOVE = "move"
    SELECT = "select"
    UPLOAD = "upload"
    DOWNLOAD = "download"
    EXTRACT = "extract"
    # Special actions
    ME = "me"
    OCR = "ocr"
    ALL_VERSIONS = "download.all_versions"
    LAST_VERSION_ONLY = "download.last_version_only"
    UPDATE_TITLE = "update.title"
    UPDATE_TAGS = "update.tags"


class Scopes:
    """Permission scopes"""

    # Node permissions
    NODE_CREATE = "node.create"
    NODE_VIEW = "node.view"
    NODE_UPDATE = "node.update"
    NODE_DELETE = "node.delete"
    NODE_MOVE = "node.move"

    # Document permissions
    DOCUMENT_UPLOAD = "document.upload"
    DOCUMENT_DOWNLOAD = "document.download"
    DOCUMENT_DOWNLOAD_ALL_VERSIONS = "document.download.all_versions"
    DOCUMENT_DOWNLOAD_LAST_VERSION_ONLY = "document.download.last_version_only"
    DOCUMENT_UPDATE_TITLE = "document.update.title"
    DOCUMENT_UPDATE_TAGS = "document.update.tags"

    # Tag permissions
    TAG_SELECT = "tag.select"  # User can assign existing tags from the catalog
    TAG_CREATE = "tag.create"
    TAG_VIEW = "tag.view"
    TAG_UPDATE = "tag.update"
    TAG_DELETE = "tag.delete"

    # User permissions
    USER_CREATE = "user.create"
    USER_SELECT = "user.select"  # User can pick from dropdown
    USER_VIEW = "user.view"
    USER_UPDATE = "user.update"
    USER_DELETE = "user.delete"
    USER_ME = "user.me"

    # Group permissions
    GROUP_CREATE = "group.create"
    GROUP_SELECT = "group.select"  # User can pick from dropdown
    GROUP_VIEW = "group.view"
    GROUP_UPDATE = "group.update"
    GROUP_DELETE = "group.delete"

    # Role permissions
    ROLE_CREATE = "role.create"
    ROLE_VIEW = "role.view"
    ROLE_SELECT = "role.select"  # User can pick from dropdown
    ROLE_UPDATE = "role.update"
    ROLE_DELETE = "role.delete"

    # Task permissions
    TASK_OCR = "task.ocr"

    # OCR language permissions
    OCRLANG_VIEW = "ocrlang.view"

    # Page permissions
    PAGE_VIEW = "page.view"      # absolete: will be removed in 3.6 (or 3.7)
    PAGE_UPDATE = "page.update"  # absolete: will be removed in 3.6 (or 3.7)
    PAGE_REORDER = "page.reorder"
    PAGE_ROTATE = "page.rotate"
    PAGE_MOVE = "page.move"
    PAGE_EXTRACT = "page.extract"
    PAGE_DELETE = "page.delete"

    # Shared node permissions
    SHARED_NODE_CREATE = "shared_node.create"
    SHARED_NODE_VIEW = "shared_node.view"
    SHARED_NODE_UPDATE = "shared_node.update"
    SHARED_NODE_DELETE = "shared_node.delete"

    # Legal portal (shared catalog under portal_settings.portal_root_node_id)
    PORTAL_VIEW = "portal.view"
    PORTAL_FEED_VIEW = "portal.feed.view"
    PORTAL_FEED_MANAGE = "portal.feed.manage"
    PORTAL_SECTION_CREATE = "portal.section.create"
    PORTAL_SECTION_UPDATE = "portal.section.update"
    PORTAL_SECTION_DELETE = "portal.section.delete"
    PORTAL_DOCUMENT_UPLOAD = "portal.document.upload"
    PORTAL_DOCUMENT_UPDATE = "portal.document.update"
    PORTAL_DOCUMENT_DELETE = "portal.document.delete"

    # Comment permissions (document library / collaboration).
    # Users may always edit/delete their own comments; update/delete scopes
    # grant moderation over other users' comments.
    COMMENT_CREATE = "comment.create"
    COMMENT_UPDATE = "comment.update"
    COMMENT_DELETE = "comment.delete"

    @classmethod
    def all_scopes(cls) -> Set[str]:
        """Return all available scopes."""
        return {
            value for name, value in cls.__dict__.items()
            if isinstance(value, str) and not name.startswith('_')
        }

    @classmethod
    def get_scopes_by_category(cls, category: str) -> List[str]:
        """Get all scopes for a specific category."""
        return [
            value for name, value in cls.__dict__.items()
            if isinstance(value, str) and value.startswith(f"{category}.")
        ]

    @classmethod
    def get_crud_scopes(cls, category: str) -> Dict[str, str]:
        """Get CRUD scopes for a category."""
        crud_actions = ["create", "view", "update", "delete"]
        result = {}

        for action in crud_actions:
            scope_name = f"{category.upper()}_{action.upper()}"
            if hasattr(cls, scope_name):
                result[action] = getattr(cls, scope_name)

        return result


# Legacy dictionary for backward compatibility
# This maintains the original SCOPES dictionary structure
SCOPES = {scope: scope for scope in Scopes.all_scopes()}

# Export all scope constants at module level for backward compatibility
# This allows imports like: from papermerge.core.features.auth import scopes
# And usage like: scopes.NODE_VIEW, scopes.USER_CREATE, etc.
_scope_attrs = [attr for attr in dir(Scopes) if not attr.startswith('_') and not callable(getattr(Scopes, attr))]
for _attr in _scope_attrs:
    globals()[_attr] = getattr(Scopes, _attr)

File: frontend/apps/ui/src/scopes.ts
export const NODE_CREATE = "node.create"
export const NODE_VIEW = "node.view"
export const NODE_UPDATE = "node.update"
export const NODE_DELETE = "node.delete"
export const NODE_MOVE = "node.move"
export const COMMANDER_VIEW = "commander.view"
export const PORTAL_VIEW = "portal.view"
export const PORTAL_FEED_VIEW = "portal.feed.view"
export const PORTAL_FEED_MANAGE = "portal.feed.manage"
export const PORTAL_SECTION_CREATE = "portal.section.create"
export const PORTAL_SECTION_UPDATE = "portal.section.update"
export const PORTAL_SECTION_DELETE = "portal.section.delete"
export const PORTAL_DOCUMENT_UPLOAD = "portal.document.upload"
export const PORTAL_DOCUMENT_UPDATE = "portal.document.update"
export const PORTAL_DOCUMENT_DELETE = "portal.document.delete"
export const CITIZEN_CATEGORY_VIEW = "citizen_category.view"
export const CITIZEN_CATEGORY_CREATE = "citizen_category.create"
export const CITIZEN_CATEGORY_UPDATE = "citizen_category.update"
export const CITIZEN_CATEGORY_DELETE = "citizen_category.delete"
export const COMMENT_CREATE = "comment.create"
export const COMMENT_UPDATE = "comment.update"
export const COMMENT_DELETE = "comment.delete"
export const DOCUMENT_UPDATE_TAGS = "document.update.tags"
export const DOCUMENT_FULL_VERSION_VIEW = "document.full_version.view"
export const DOCUMENT_FULL_VERSION_MANAGE = "document.full_version.manage"
export const DOCUMENT_UPLOAD = "document.upload"
export const DOCUMENT_DOWNLOAD = "document.download"
export const TAG_CREATE = "tag.create"
export const TAG_VIEW = "tag.view"
export const TAG_SELECT = "tag.select"
export const TAG_UPDATE = "tag.update"
export const TAG_DELETE = "tag.delete"
export const USER_CREATE = "user.create"
export const USER_VIEW = "user.view"
export const USER_UPDATE = "user.update"
export const USER_DELETE = "user.delete"
export const USER_ME = "user.me"
export const USER_SELECT = "user.select"
export const GROUP_CREATE = "group.create"
export const GROUP_VIEW = "group.view"
export const GROUP_SELECT = "group.select"
export const GROUP_UPDATE = "group.update"
export const GROUP_DELETE = "group.delete"
export const ROLE_CREATE = "role.create"
export const ROLE_VIEW = "role.view"
export const ROLE_SELECT = "role.select"
export const ROLE_UPDATE = "role.update"
export const ROLE_DELETE = "role.delete"
export const TASK_OCR = "task.ocr"
export const OCRLANG_VIEW = "ocrlang.view"
export const PAGE_VIEW = "page.view"
export const PAGE_UPDATE = "page.update"
export const PAGE_MOVE = "page.move"
export const PAGE_EXTRACT = "page.extract"
export const PAGE_DELETE = "page.delete"

/** Tag admin: create / update / delete on /tags (moderator & admin by default). */
export function canManageTags(scopes: string[]): boolean {
  return (
    scopes.includes(TAG_CREATE) ||
    scopes.includes(TAG_UPDATE) ||
    scopes.includes(TAG_DELETE)
  )
}

/** Assign tags on documents/folders (``document.update.tags`` or full ``node.update``). */
export function canUpdateNodeTags(scopes: string[]): boolean {
  return scopes.includes(NODE_UPDATE) || scopes.includes(DOCUMENT_UPDATE_TAGS)
}

/** Pick existing tags from the catalog and assign them to a node. */
export function canAssignNodeTags(scopes: string[]): boolean {
  return scopes.includes(TAG_SELECT) || canUpdateNodeTags(scopes)
}

/** Pick tags from the organization catalog when editing a node. */
export function canSelectTags(scopes: string[]): boolean {
  return scopes.includes(TAG_SELECT) || canUpdateNodeTags(scopes)
}

export type CommanderWriteContext = {
  /** Folder is under the legal portal catalog and route is ``/folder/…``. */
  isPortalContext: boolean
}

/** Open the file manager UI (``/folder/…``). */
export function canOpenCommander(scopes: string[]): boolean {
  return scopes.includes(COMMANDER_VIEW)
}

export function canUploadInCommander(
  scopes: string[],
  ctx: CommanderWriteContext
): boolean {
  if (ctx.isPortalContext) {
    return scopes.includes(PORTAL_DOCUMENT_UPLOAD)
  }
  return scopes.includes(DOCUMENT_UPLOAD)
}

export function canCreateFolderInCommander(
  scopes: string[],
  ctx: CommanderWriteContext
): boolean {
  if (ctx.isPortalContext) {
    return scopes.includes(PORTAL_SECTION_CREATE)
  }
  return scopes.includes(NODE_CREATE)
}

export function canDeleteInCommander(
  scopes: string[],
  ctx: CommanderWriteContext
): boolean {
  if (ctx.isPortalContext) {
    return (
      scopes.includes(PORTAL_SECTION_DELETE) ||
      scopes.includes(PORTAL_DOCUMENT_DELETE)
    )
  }
  return scopes.includes(NODE_DELETE)
}

export function canRenameInCommander(
  scopes: string[],
  ctx: CommanderWriteContext
): boolean {
  if (ctx.isPortalContext) {
    return (
      scopes.includes(PORTAL_SECTION_UPDATE) ||
      scopes.includes(PORTAL_DOCUMENT_UPDATE)
    )
  }
  return scopes.includes(NODE_UPDATE)
}

export function canReorderInCommander(
  scopes: string[],
  ctx: CommanderWriteContext
): boolean {
  return canRenameInCommander(scopes, ctx)
}

export function canMoveInCommander(
  scopes: string[],
  ctx: CommanderWriteContext
): boolean {
  if (ctx.isPortalContext) {
    return (
      scopes.includes(PORTAL_SECTION_UPDATE) ||
      scopes.includes(PORTAL_DOCUMENT_UPDATE)
    )
  }
  return scopes.includes(NODE_MOVE)
}

export function canUseMspTemplateInCommander(scopes: string[]): boolean {
  return scopes.includes(NODE_CREATE)
}

export function canMutateInCommander(
  scopes: string[],
  ctx: CommanderWriteContext
): boolean {
  return (
    canUploadInCommander(scopes, ctx) ||
    canCreateFolderInCommander(scopes, ctx) ||
    canDeleteInCommander(scopes, ctx) ||
    canRenameInCommander(scopes, ctx) ||
    canMoveInCommander(scopes, ctx)
  )
}

export const ALL_PERMS = [
  DOCUMENT_DOWNLOAD,
  DOCUMENT_UPLOAD,
  PAGE_VIEW,
  PAGE_MOVE,
  PAGE_UPDATE,
  PAGE_DELETE,
  PAGE_EXTRACT,
  USER_VIEW,
  USER_CREATE,
  USER_UPDATE,
  USER_DELETE,
  USER_ME,
  GROUP_VIEW,
  GROUP_CREATE,
  GROUP_UPDATE,
  GROUP_DELETE,
  ROLE_VIEW,
  ROLE_CREATE,
  ROLE_UPDATE,
  ROLE_DELETE,
  TAG_VIEW,
  TAG_SELECT,
  TAG_CREATE,
  TAG_UPDATE,
  TAG_DELETE,
  NODE_VIEW,
  COMMANDER_VIEW,
  NODE_MOVE,
  NODE_CREATE,
  NODE_UPDATE,
  DOCUMENT_UPDATE_TAGS,
  DOCUMENT_FULL_VERSION_VIEW,
  DOCUMENT_FULL_VERSION_MANAGE,
  NODE_DELETE,
  COMMENT_CREATE,
  COMMENT_UPDATE,
  COMMENT_DELETE,
  TASK_OCR,
  OCRLANG_VIEW,
  PORTAL_VIEW,
  PORTAL_FEED_VIEW,
  PORTAL_FEED_MANAGE,
  PORTAL_SECTION_CREATE,
  PORTAL_SECTION_UPDATE,
  PORTAL_SECTION_DELETE,
  PORTAL_DOCUMENT_UPLOAD,
  PORTAL_DOCUMENT_UPDATE,
  PORTAL_DOCUMENT_DELETE,
  CITIZEN_CATEGORY_VIEW,
  CITIZEN_CATEGORY_CREATE,
  CITIZEN_CATEGORY_UPDATE,
  CITIZEN_CATEGORY_DELETE
]
